gdal/port/cpl_userfaultfd.cpp

```diff
@@ -0,0 +1,214 @@
+    offset = reinterpret_cast<uint64_t>(fault_addr) - reinterpret_cast<uint64_t>(ctx->vma_ptr);
+
+    // Copy data into page
+    VSIFSeekL(file, offset, SEEK_SET);
+    VSIFReadL(ctx->page_ptr, ctx->page_size, 1, file);
+
+    uffdio_copy.src = reinterpret_cast<uintptr_t>(ctx->page_ptr);
+    uffdio_copy.dst = fault_addr;
+    uffdio_copy.len = static_cast<uintptr_t>(ctx->page_size);
+    uffdio_copy.mode = 0;
+    uffdio_copy.copy = 0;
+    if (ioctl(ctx->uffd, UFFDIO_COPY, &uffdio_copy) == -1) break;
+  }
+
+  // Return resources
+  VSIFCloseL(file);
+
+  pthread_cleanup_pop(1);
+  return nullptr;
+}
+
+/*
+ * Returns nullptr on failure, a valid pointer on success.
+ */
+void * hookup(const char * pszFilename, void ** ppVma, uint64_t * pnVmaSize)
+{
+  VSIStatBufL statbuf;
+  pthread_t thread;
+  struct cpl_uffd_context * ctx = nullptr;
+  struct uffdio_api uffdio_api;
+
+  // Get the size of the asset
+  if (VSIStatL(pszFilename, &statbuf)) return nullptr;
+
+  // Setup the `cpl_uffd_context` struct
+  ctx = static_cast<struct cpl_uffd_context *>(malloc(sizeof(struct cpl_uffd_context)));
+  strncpy(ctx->filename, pszFilename, sizeof((reinterpret_cast<struct cpl_uffd_context *>(0))->filename));
+  ctx->page_size = sysconf(_SC_PAGESIZE);
+  ctx->vma_size = ((statbuf.st_size/ctx->page_size)+1) * ctx->page_size;
+
+  // If the mmap failed, free resources and return
+  ctx->vma_ptr = mmap(nullptr, ctx->vma_size, PROT_READ, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0);
+  if (ctx->vma_ptr == BAD_MMAP) {
+    ctx->vma_ptr = nullptr;
+    uffd_cleanup(ctx);
+    return nullptr;
+  }
+
+  // Attempt to aquire a scratch page to use to fulfill requests.
+  ctx->page_ptr = mmap(nullptr, ctx->page_size, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0);
+  if (ctx->page_ptr == BAD_MMAP) {
+    ctx->page_ptr = nullptr;
+    uffd_cleanup(ctx);
+    return nullptr;
+  }
+
+  // Get non-blocking file descriptor
+  if ((ctx->uffd = syscall(__NR_userfaultfd, O_CLOEXEC | O_NONBLOCK)) == -1) {
+    ctx->uffd = -1;
+    uffd_cleanup(ctx);
+    return nullptr;
+  }
+
+  // Query API
+  uffdio_api.api = UFFD_API;
+  uffdio_api.features = 0;
+  if (ioctl(ctx->uffd, UFFDIO_API, &uffdio_api) == -1) {
+    uffd_cleanup(ctx);
+    return nullptr;
+  }
+
+  // Register memory range
+  ctx->uffdio_register.range.start = reinterpret_cast<uintptr_t>(ctx->vma_ptr);
+  ctx->uffdio_register.range.len = ctx->vma_size;
+  ctx->uffdio_register.mode = UFFDIO_REGISTER_MODE_MISSING;
+  if (ioctl(ctx->uffd, UFFDIO_REGISTER, &ctx->uffdio_register) == -1) {
+    uffd_cleanup(ctx);
+    return nullptr;
+  }
+
+  // Start handler
+  if (pthread_create(&thread, nullptr, fault_handler, ctx)) {
+    uffd_cleanup(ctx);
+    return nullptr;
+  }
+
+  *ppVma = ctx->vma_ptr;
+  *pnVmaSize = ctx->vma_size;
+  return ctx;
+}
+
+void unhook(void * ptr)
+{
+  struct cpl_uffd_context * ctx = static_cast<struct cpl_uffd_context *>(ptr);
+  if (ctx && ctx->uffd != -1) close(ctx->uffd);
```

Comment:
ctx itself should be freed. Actually, shoulnd't uffd_cleanup() itself be called

Reply:
Here, I assume that the thread is still running or has run.  If that is the case, then the cleanup hook will cause `uffd_cleanup` to be called.  (The assumption is that `unhook` is never called unless the thread has successfully started and that the thread cannot end without its cleanup hook firing.)

Reply:
ok, could be worth explaining that as a comment.

Reply:
Okay, added.